File: python/array/SpiralMatrix.py
class Solution:
    # @param matrix, a list of lists of integers
    # @return a list of integers
    def spiralOrder(self, matrix):
        rows = len(matrix)
        if rows == 0:
            return []
        cols = len(matrix[0])
        total = rows * cols
        direction = 0
        col = 0
        row = 0
        i = 0
        startRow = 0
        startCol = 0
        result = []
        while i < total:
            i += 1
            result.append(matrix[row][col])
            if direction == 0:
                if col == cols - 1:
                    direction = 1
                    row += 1
                else:
                    col += 1
            elif direction == 1:
                if row == rows - 1:
                    direction = 2
                    col -= 1
                else:
                    row += 1
            elif direction == 2:
                if col == startCol:
                    row -= 1
                    direction = 3
                else:
                    col -= 1
            else:
                if row == startRow + 1:
                    startRow += 1
                    startCol += 1
                    cols -= 1
                    rows -= 1
                    col += 1
                    direction = 0
                else:
                    row -= 1
                    
        return result
                

if __name__ == "__main__":
    solution = Solution()

    print solution.spiralOrder([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 16]])

File: archive/cpp/array/SpiralMatrix.cpp
#include <vector>

class Solution {
public:
    std::vector<int> spiralOrder(std::vector<std::vector<int> > &matrix) {
        std::vector<int> result;
        int rows = int(matrix.size());
        if (rows == 0) {
            return result;
        }

        int cols = int(matrix[0].size());
        int total = rows * cols;
        int i = 0;
        int startRow = 0;
        int startCol = 0;
        int direction = 0;
        int row = 0;
        int col = 0;
        while (i < total) {
            i ++;
            result.push_back(matrix[row][col]);
            if (direction == 0) {
                if (col == cols - 1) {
                    row ++;
                    direction = 1;
                } else {
                    col ++;
                }
            } else if (direction == 1) {
                if (row == rows - 1) {
                    col --;
                    direction = 2;
                } else {
                    row ++;
                }
            } else if (direction == 2) {
                if (col == startCol) {
                    row --;
                    direction = 3;
                } else {
                    col --;
                }
            } else {
                if (row == startRow - 1) {
                    col ++;
                    startRow ++;
                    startCol ++;
                    cols --;
                    rows --;
                    direction = 1;
                } else {
                    row --;
                }
            }
        }
        return result;
    }
};
